flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericWriteAheadSink.java

```diff
@@ -77,111 +84,144 @@ public GenericWriteAheadSink(CheckpointCommitter committer, TypeSerializer<IN> s
+
+			PendingCheckpoint pendingCheckpoint = new PendingCheckpoint(
+				checkpointId, subtaskIdx, timestamp, handle);
+
+			if (pendingCheckpoints.contains(pendingCheckpoint)) {
 				//we already have a checkpoint stored for that ID that may have been partially written,
 				//so we discard this "alternate version" and use the stored checkpoint
 				handle.discardState();
 			} else {
-				state.pendingHandles.put(checkpointId, new Tuple2<>(timestamp, handle));
+				pendingCheckpoints.add(pendingCheckpoint);
 			}
 			out = null;
 		}
 	}
 
 	@Override
-	public void snapshotState(FSDataOutputStream out,
-			long checkpointId,
-			long timestamp) throws Exception {
+	public void snapshotState(FSDataOutputStream out, long checkpointId, long timestamp) throws Exception {
 		saveHandleInState(checkpointId, timestamp);
 
-		InstantiationUtil.serializeObject(out, state);
+		DataOutputViewStreamWrapper outStream = new DataOutputViewStreamWrapper(out);
+		outStream.writeInt(pendingCheckpoints.size());
+		for (PendingCheckpoint pendingCheckpoint : pendingCheckpoints) {
+			pendingCheckpoint.serialize(outStream);
+		}
 	}
 
 	@Override
 	public void restoreState(FSDataInputStream in) throws Exception {
-		this.state = InstantiationUtil.deserializeObject(in, getUserCodeClassloader());
+		final DataInputViewStreamWrapper inStream = new DataInputViewStreamWrapper(in);
+		int numPendingHandles = inStream.readInt();
+		for (int i = 0; i < numPendingHandles; i++) {
+			pendingCheckpoints.add(PendingCheckpoint.restore(inStream, getUserCodeClassloader()));
+		}
 	}
 
-	private void cleanState() throws Exception {
-		synchronized (this.state.pendingHandles) { //remove all handles that were already committed
-			Set<Long> pastCheckpointIds = this.state.pendingHandles.keySet();
-			Set<Long> checkpointsToRemove = new HashSet<>();
-			for (Long pastCheckpointId : pastCheckpointIds) {
-				if (committer.isCheckpointCommitted(pastCheckpointId)) {
-					checkpointsToRemove.add(pastCheckpointId);
+	/**
+	 * Called at {@link #open()} to clean-up the pending handle list.
+	 * It iterates over all restored pending handles, checks which ones are already
+	 * committed to the outside storage system and removes them from the list.
+	 */
+	private void cleanRestoredHandles() throws Exception {
+		synchronized (pendingCheckpoints) {
+
+			// for each of the pending handles...
+			Iterator<PendingCheckpoint> pendingCheckpointIt = pendingCheckpoints.iterator();
+			while (pendingCheckpointIt.hasNext()) {
+
+				PendingCheckpoint pendingCheckpoint = pendingCheckpointIt.next();
+				long checkpointId = pendingCheckpoint.checkpointId;
+				int subtaskId = pendingCheckpoint.subtaskId;
+
+				//...check if the temporary buffer is already committed and if yes,
+				// remove it from the list of pending checkpoints.
+				if (committer.isCheckpointCommitted(subtaskId, checkpointId)) {
+					pendingCheckpoint.stateHandle.discardState();
+					pendingCheckpointIt.remove();
```

Comment:
This will throw a `ConcurrentModificationException` according to the [javadocs](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html).